### ISSUE TITLE
XML validation throws warning at PMDException

### ISSUE THREAD

**Issue by robhva** [forcedotcom/sfdx-scanner#277]
Hi,

I am trying to setup XML validation. To do so I have customised the file `Config.json` and added xml to `supportedLanguages` but also removed the exclamation-mark from `**/*-meta.xml` within `targetPatterns`.

I am not persuing advanced validation I only want to know if the layout is correct (for example do all open tags have equivalent close tags etc...). I am expecting the default xml library in PMD should be more than capable to support this simple validation.

I created a malformed xml (added an additional opening tag without a closing one) and run SFDX-scanner using:
`sfdx scanner:run --category="Error Prone" --violations-cause-error --target "force-app/main/default"`

This is working to some extend, however A warning is thrown instead of an exception. While as the warning text states "Message: PMD**Exception**: Error while parsing". Which is a bit strange as xml files which can not be parsed can also not be deployed onto a salesforce environment. As such I am expecting this to be properly supported. I also do not see an indication where the issue is caused (for example the line number ).

If I do the same using native PMD, I get the result:
`[Fatal Error] :25:7: The element type "values" must be terminated by the matching end-tag "</values>".`
`✔ PMD analysis was successful.`

As you can see the behaviour is identical. Successful analysis.
Hoever I did set the parameter `failOnViolation: true`.

Is there something what can be done to have an exception reported as a violation?

Regards,
Rob

**Comment by robhva** [comment]
Is anyone looking at this request?

As additional information: 

If a malformed xml file is deployed to Salesforce it will generate an error message like the one below: 
```
Error parsing file: LWC Metadata Xml Parser: ParseError at [row,col]:[17,3] 
Message: The element type "targets" must be terminated by the matching end-tag "</targets>". 
XML parse error: Unexpected element: targets 
```

Basically I am expecting sfdx-scanner to be capable to trow the same kind of error message. 
Instead it currently only displays a warning using less information. 
It only tells you: "PMD failed to evaluate against file" 

Can you please apply the same xml file validation logic, as being applied while deploying a file to Salesforce, into sfdx scanner?

This will help us to pushback developers early with useful information in a DevOps pipeline.

Thank you in advance!

**Comment by jfeingold35** [comment]
Hey @robhva. Sorry for the wait time. Just to be clear I understand what you're asking for, you'd like for the scanner to provide more detailed information about parsing failures?

**Comment by robhva** [comment]
Hi Josh,

Thank you for your reply. I also read your comment at GitHub.
Basically we can have this split into two requirements:
1. Report the issue as a violation as a bare minimum as this will make is
possible to enable the parameter: "--violations-cause-error"
2. Provide more detailed information about parsing failures.

Within a DevOps pipeline sfdx-scanner is used to quickly validate the
content to be used/deployed. We are using the sfdx-scanner report and
attach it as an artefact to our pipeline so that developers have a "One
Stop Shop" with regards to their development activities (we do the same for
regression and deployment tests). sfdx-scanner at this point becomes
useless for this purpose as it does not 'break' the pipeline. As such also
a deployment test is started (so that we are able to guarantee a successful
deployment) and this breaks instead. As a result pipeline run-time is spent
while as we could have been able to catch the error in the first place by
sfdx-scanner.

Like expressed within my second comment it turns-out that Salesforce has a
wonderful logic in place to do exactly what I am asking for. It throws a
warning and point clearly where a developer should be looking to have the
issue resolved. Can we have this exact same logic implement implemented in
sfdx-scanner ?

Our DevOps pipeline does more than only a deployment of a bit of code. So
each improvement we can make to guarantee a successful deployment is a win.

I hope I was able to make myself clear.
Please do not hesitate to reach out if you have additional questions.

Thank you in advance

**Comment by robhva** [comment]
Hi Josh,

In addition to your comment:
> _surface them to the user as warnings, since that's basically what they are_

I think you are making a mistake. PMD might have good reasons to indicate them as a warning. However sfdx-scanner serves a different purpose. I can not deploy a malformed xml file to Salesforce. This is also the reason why this issue is reported as an ERROR during deployment.

As mentioned before using sfdx-scanner only requires 1/5th of the time in comparison to a delta deployment. As such from a Salesforce DevOps pipeline perspective there is much to gain if we can properly exit the scanner for issues which can not be deployed in the first place.

Rob

**Comment by jfeingold35** [comment]
That's a good point that I hadn't considered before. The team is going to look at converting these parse failures into proper violations and providing a consistent experience across all engines.

**Comment by ranekere-sfdc** [comment]
The fix for this issue will be released as part of Scanner 2.5

**Comment by robhva** [comment]
Just to confirm it's working as expected. Thanks!
